File: 3805-CountCaesarCipherPairs/3805-CountCaesarCipherPairs.py
# Last updated: 07/09/2026, 22:23:19
class Solution:
    def countPairs(self, words: List[str]) -> int:
        if (len(words) <= 1):
            return 0

        n = len(words)
        m = len(words[0])

        patterns = defaultdict(int)

        for word in words:
            if (m == 1):
                pattern = tuple()
            else:
                pattern = tuple((ord(word[i]) - ord(word[i - 1])) % 26 for i in range(1, m))
            patterns[pattern] += 1

        result = 0
        print (patterns)
        for count in patterns.values():
            if (count >= 2):
                result += (count * (count - 1)) // 2
        return result

File: 3805-CountCaesarCipherPairs/3805-CountCaesarCipherPairs.cpp
// Last updated: 20/09/2026, 10:29:53
class Solution {
public:
    long long countPairs(vector<string>& words) {
        if (words.size() <= 1) return 0;
        unordered_map<string, long long> mp;
        for (const string& str : words){
            string pattern = "";
            for (int i = 0; i < str.size() - 1; ++i){
                pattern += 'a' + (str[i + 1] - str[i] + 26) % 26;
            }
            ++mp[pattern];
        }
        long long result = 0;
        for (auto [m, f]: mp){
            if (f >= 2){
                result += f * (f - 1) / 2;
            }
        }
        return result;
    }
};